src/ir/eh-utils.cpp

```diff
@@ -16,53 +16,137 @@
 
 namespace wasm {
 
 namespace EHUtils {
 
-bool isPopValid(Expression* catchBody) {
-  Expression* firstChild = nullptr;
-  auto* block = catchBody->dynCast<Block>();
-  if (!block) {
-    firstChild = catchBody;
-  } else {
-    // When there are multiple expressions within a catch body, an implicit
-    // block is created within it for convenience purposes, and if there are no
-    // branches that targets the block, it will be omitted when written back.
-    // But if there is a branch targetting this block, this block cannot be
-    // removed, and 'pop''s location will be like
-    // (catch $e
-    //   (block $l0
-    //     (pop i32) ;; within a block!
-    //     (br $l0)
-    //     ...
-    //   )
-    // )
-    // which is invalid.
-    if (BranchUtils::BranchSeeker::has(block, block->name)) {
-      return false;
-    }
-    // There should be a pop somewhere
-    if (block->list.empty()) {
-      return false;
-    }
-    firstChild = *block->list.begin();
-  }
+// This returns three values, some of them as output parameters:
+// - Return value: 'pop' expression (Expression*), when there is one in
+//   first-descendant line. If there's no such pop, it returns null.
+// - isPopNested: Whether the discovered 'pop' is nested within a block
+// - popPtr: 'pop' expression's pointer (Expression**), when there is one found
+//
+// When 'catchBody' itself is a 'pop', 'pop''s pointer is null, because there is
+// no way to get the given expression's address. But that's fine because pop's
+// pointer is only necessary (in handleBlockNestedPops) to fix it up when it is
+// nested, and if 'catchBody' itself is a pop, we don't need to fix it up.
+static Expression*
+getFirstPop(Expression* catchBody, bool& isPopNested, Expression**& popPtr) {
+  Expression* firstChild = catchBody;
+  isPopNested = false;
+  popPtr = nullptr;
+  // When there are multiple expressions within a catch body, an implicit
+  // block is created within it for convenience purposes.
+  auto* implicitBlock = catchBody->dynCast<Block>();
 
   // Go down the line for the first child until we reach a leaf. A pop should be
-  // in that first-decendent line.
+  // in that first-decendant line.
+  Expression** firstChildPtr = nullptr;
   while (true) {
     if (firstChild->is<Pop>()) {
-      return true;
+      popPtr = firstChildPtr;
+      return firstChild;
     }
-    // We use ValueChildIterator in order not to go into block/loop/try/if
-    // bodies, because a pop cannot be in those control flow expressions.
-    ValueChildIterator it(firstChild);
+
+    if (Properties::isControlFlowStructure(firstChild)) {
+      // If's condition is a value child who comes before an 'if' instruction
+      // in binary, it is fine if a 'pop' is in there. We don't allow a 'pop' to
+      // be in an 'if''s then or else body because they are not
+      // first descendants.
+      if (auto* if_ = firstChild->dynCast<If>()) {
+        firstChild = if_->condition;
```

Comment:
```suggestion
      if (auto* iff = firstChild->dynCast<If>()) {
        firstChild = iff->condition;
```
(this is the convention used most elsewhere, I believe)